### PR TITLE
draft for trainable pad op

### PR DESCRIPTION
**Pull request by YongseopKim**
#12413 

ONE-DCO-1.0-Signed-off-by: Yongseop Kim <yons.kim@samsung.com>

### CODE REVIEW

**Comment by YongseopKim** [comment]
```
~/github/YongseopKim/ONE (draft/onert_train_pad ✗) ./Product/out/bin/onert_train --modelfile ./op.pad/pad.circle
Model Filename ./op.pad/pad.circle
== training parameter ==
- learning_rate   = 0.001
- batch_size      = 32
- loss_info       = {loss = mean squared error, reduction = sum over batch size}
- optimizer       = sgd
========================
E: not supported random input and expected generator
```

Surprisingly, I don't know how to load input/output on `onert_train`.

**Comment by YongseopKim** [comment]
While drafting, I realized that there are ops which don't have to do backwards.

In the case of Pad, AFAIK, pad does nothing at the backward propagation. (because padding doesn't have any params to be tranied.)

Therefore, these ops that have nothing to do at the backward propagation can be extracted as a type such as `InterfaceDoNothingAtTheBackwardProgration`. (I don't know how to name it...)

**Comment by jyoungyun** [comment]
> Surprisingly, I don't know how to load input/output on `onert_train`. 

You can make a dataset for test using `./tools/generate_datafile/tf_dataset_converter/`. This tool does not support random input generator.

**Comment by jyoungyun** [comment]
> In the case of Pad, AFAIK, pad does nothing at the backward propagation. (because padding doesn't have any params to be tranied.)

Even though there is no weight in `Pad` Op, it should implement backwarding function for backpropagation.

File: runtime/onert/backend/train/ops/PadLayer.cc [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+  _output = output;
+  memcpy(_padData, padData, sizeof(_padData));
+  _padRank = padRank;
+  _constantValueData.v = constantValueData;
+  _back_prop_input = back_prop_input;
+  _back_prop_output = back_prop_output;
+}
+
+void PadLayer::forward(bool)
+{
```
Comment (jyoungyun) [inline runtime/onert/backend/train/ops/PadLayer.cc:67]: How about using cker/cpu/Pad directly instead of implementing it here?

File: compute/cker/include/cker/Types.h [hunk 2]
```diff
@@ -397,6 +397,12 @@ struct LeakyReluParams
   float alpha;
 };
 
+struct PadParams
```
Comment (YongseopKim) [inline compute/cker/include/cker/Types.h:400]: #12522

File: compute/cker/include/cker/train/operation/Pad.h [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+ * input_data(backward_output_data) will be transformed by backward of PAD operation with padding options to output_data(backward_input_data)
+ *
+ * input_data(backward_output_data)  ->  output_data(backward_input_data)
+ *   [C,C,C,C]                       ->    [0,1]
+ *   [C,0,1,C]                       ->    [2,3]
+ *   [C,2,3,C]                       ->
+ *   [C,C,C,C]                       ->
+ */
+template <typename T>
+inline void Depad(const int32_t *padding_data, int32_t pad_rank, const Shape &input_shape,
```
Comment (YongseopKim) [inline compute/cker/include/cker/train/operation/Pad.h:57]: #12522

File: compute/cker/src/train/Pad.test.cc [hunk 4]
```diff
@@ -0,0 +1,319 @@
...
+#include <cker/Shape.h>
+
+#include <gtest/gtest.h>
+#include <vector>
+
+namespace
+{
+using namespace nnfw::cker;
+
+template <typename T> class PadOpVerifier
```
Comment (YongseopKim) [inline compute/cker/src/train/Pad.test.cc:28]: #12522

File: compute/cker/src/train/Pad.test.cc [hunk 5]
```diff
@@ -0,0 +1,319 @@
...
+    if (expect_eq)
+      EXPECT_EQ(expected_backward_input, backward_input);
+    else
+      EXPECT_NE(expected_backward_input, backward_input);
+  }
+};
+
+} // namespace
+
+TEST(CKer_Operation, Pad)
```
Comment (YongseopKim) [inline compute/cker/src/train/Pad.test.cc:83]: #12522

File: runtime/onert/backend/train/ops/PadLayer.cc (hunk 1 above)
Comment (YongseopKim) [reply runtime/onert/backend/train/ops/PadLayer.cc:67]: Maybe you mentioned on #12522, right?

File: runtime/onert/core/src/ir/train/operation/Pad.cc [hunk 6]
```diff
@@ -0,0 +1,49 @@
+/*
```
Comment (YongseopKim) [inline runtime/onert/core/src/ir/train/operation/Pad.cc:1]: https://github.com/Samsung/ONE/pull/12524

File: runtime/onert/core/src/compiler/train/TrainableOperationConverter.h [hunk 7]
```diff
@@ -42,6 +42,7 @@ class TrainableOperationConverter : public UntrainableOperationConverter
   void visit(const ir::operation::ElementwiseActivation &) override;
   void visit(const ir::operation::FullyConnected &) override;
   void visit(const ir::operation::Loss &node) override;
+  void visit(const ir::operation::Pad &node) override;
```
Comment (YongseopKim) [inline runtime/onert/core/src/compiler/train/TrainableOperationConverter.h:45]: https://github.com/Samsung/ONE/pull/12524

File: runtime/onert/core/include/ir/train/operation/Pad.h [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+namespace onert
+{
+namespace ir
+{
+namespace train
+{
+namespace operation
+{
+
+class Pad : public ir::operation::Pad, public ITrainableOperation
```
Comment (YongseopKim) [inline runtime/onert/core/include/ir/train/operation/Pad.h:32]: https://github.com/Samsung/ONE/pull/12524

File: runtime/onert/core/include/ir/train/Operations.lst [hunk 9]
```diff
@@ -23,6 +23,7 @@ OP(DepthwiseConv2D)
 OP(ElementwiseActivation)
 OP(FullyConnected)
 OP(Loss)
+OP(Pad)
```
Comment (YongseopKim) [inline runtime/onert/core/include/ir/train/Operations.lst:26]: https://github.com/Samsung/ONE/pull/12524

File: runtime/onert/core/include/ir/train/Operations.Include.h [hunk 10]
```diff
@@ -22,6 +22,7 @@
 #include "ir/train/operation/ElementwiseActivation.h"
 #include "ir/train/operation/FullyConnected.h"
 #include "ir/train/operation/Loss.h"
+#include "ir/train/operation/Pad.h"
```
Comment (YongseopKim) [inline runtime/onert/core/include/ir/train/Operations.Include.h:25]: https://github.com/Samsung/ONE/pull/12524

File: runtime/onert/core/src/compiler/train/TrainableOperationConverter.h (hunk 7 above)
Comment (YongseopKim) [inline runtime/onert/core/src/compiler/train/TrainableOperationConverter.h:45]: https://github.com/Samsung/ONE/pull/12534

File: runtime/onert/core/src/compiler/train/TrainableOperationConverter.cc [hunk 11]
```diff
@@ -66,6 +66,11 @@ void TrainableOperationConverter::visit(const ir::operation::Loss &node)
   _return_op = std::make_unique<ir::train::operation::Loss>(node, _training_info->lossInfo());
 }
 
+void TrainableOperationConverter::visit(const ir::operation::Pad &node)
```
Comment (YongseopKim) [inline runtime/onert/core/src/compiler/train/TrainableOperationConverter.cc:69]: https://github.com/Samsung/ONE/pull/12534

File: runtime/onert/backend/train/ops/OperationUtils.h [hunk 12]
```diff
@@ -35,6 +35,7 @@ using cpu::ops::getShape;
 using cpu::ops::getNumberOfDimensions;
 using cpu::ops::getNumberOfElements;
 using cpu::ops::getSizeOfDimension;
+using cpu::ops::ConstDataPtr;
```
Comment (YongseopKim) [inline runtime/onert/backend/train/ops/OperationUtils.h:38]: https://github.com/Samsung/ONE/pull/12535

File: runtime/onert/backend/train/ops/PadLayer.cc [hunk 13]
```diff
@@ -0,0 +1,121 @@
+/*
```
Comment (YongseopKim) [inline runtime/onert/backend/train/ops/PadLayer.cc:1]: https://github.com/Samsung/ONE/pull/12535

File: runtime/onert/backend/train/ops/PadLayer.h [hunk 14]
```diff
@@ -0,0 +1,68 @@
+/*
```
Comment (YongseopKim) [inline runtime/onert/backend/train/ops/PadLayer.h:1]: https://github.com/Samsung/ONE/pull/12535

**Comment by YongseopKim** [comment]
Maybe last PR: https://github.com/YongseopKim/ONE/tree/pr/train/pad/KernelGenerator

File: runtime/onert/backend/train/KernelGenerator.cc [hunk 15]
```diff
@@ -270,6 +271,37 @@ void KernelGenerator::visit(const ir::train::operation::Loss &node)
   }
 }
 
+void KernelGenerator::visit(const ir::train::operation::Pad &node)
```
Comment (YongseopKim) [inline runtime/onert/backend/train/KernelGenerator.cc:274]: https://github.com/Samsung/ONE/pull/12555

File: runtime/onert/backend/train/KernelGenerator.h [hunk 16]
```diff
@@ -50,6 +50,7 @@ class KernelGenerator : public backend::train::KernelGeneratorBase
   void visit(const ir::train::operation::ElementwiseActivation &) override;
   void visit(const ir::train::operation::FullyConnected &) override;
   void visit(const ir::train::operation::Loss &) override;
+  void visit(const ir::train::operation::Pad &) override;
```
Comment (YongseopKim) [inline runtime/onert/backend/train/KernelGenerator.h:53]: https://github.com/Samsung/ONE/pull/12555
